roofit/roofitcore/inc/RooFit/Detail/CodeSquashContext.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Project: RooFit
+ * Authors:
+ *   Garima Singh, CERN 2023
+ *   Jonas Rembser, CERN 2023
+ *
+ * Copyright (c) 2023, CERN
+ *
+ * Redistribution and use in source and binary forms,
+ * with or without modification, are permitted according to the terms
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)
+ */
+
+#ifndef RooFit_Detail_CodeSquashContext_h
+#define RooFit_Detail_CodeSquashContext_h
+
+#include <string>
+#include <unordered_map>
+
+namespace RooFit {
+
+namespace Detail {
+
+/// @brief A class to maintain the context for squashing of RooFit models into code.
+class CodeSquashContext {
+   /// @brief Map of node names to their result strings.
+   std::unordered_map<std::string, std::string> nodeNames;
+   /// @brief Block of code that is placed before the rest of the function body.
+   std::string globalScope;
+
+public:
+   /// @brief Adds (or overwrites) the string representing the result of a node (aka "node result name").
+   /// @param key The name of the node to get the result name for.
+   /// @param value The new name to assign/overwrite.
+   inline void addResultName(std::string const &key, std::string value) { nodeNames[key] = value; }
+
+   /// @brief Gets the result name for the given node using the node name.
+   /// @param key Name of the node to get the result string for.
+   /// @return String representing the result of this node.
+   inline std::string getResultName(std::string const &key) { return nodeNames[key]; }
+
+   /// @brief Checks if the current node has a result string already assigned.
+   /// @param key Name of the node to get the result string for.
+   /// @return True if the node was assigned a result string, false otherwise.
+   inline bool isResultNameAssigned(std::string const &key) { return nodeNames.find(key) != nodeNames.end(); }
+
+   /// @brief Adds the given string to the string block that will be emitted at the top of the squashed function. Useful
+   /// for variable declarations.
+   /// @param str The string to add to the global scope.
+   inline void addToGlobalScope(std::string str) { globalScope += str; }
+
+   /// @brief Assemble the final code with the return expression and global statements.
+   /// @param returnExpr The string representation of what the squashed function should return, usually the head node.
+   /// @param outCodeBody The body of the function, also the final output of the function.
+   inline void assembleCode(std::string returnExpr, std::string &outCodeBody)
```

Comment:
```suggestion
   inline void assembleCode(std::string const& returnExpr, std::string &outCodeBody)
```
Avoid unnecessary string copy

Reply:
Actually, this (and the last ones) are deliberately kept non-cont to avoid extra overhead because I read [here](url) that It's okay to pass by value in cases where the string is copied and also that it's okay to return the string by value as more often that not it is copy elision'd.

With that being said, I know a lot of people also advise against it. So I'll leave it up your experience - i.e. if you want me to change it I will.

Reply:
I don't think it is just limited to temporaries, but okay.